src/java.base/share/classes/jdk/internal/ref/CleanerImpl.java

```diff
@@ -231,4 +224,78 @@ protected void performCleanup() {
             // no action
         }
     }
+
+    /**
+     * A specialized implementation that tracks phantom cleanables.
+     * Backing storage is expanded and trimmed automatically.
+     * Insert/remove run in amortized constant time.
+     */
+    static final class PhantomCleanableList {
+        private final ArrayList<PhantomCleanable<?>> list = new ArrayList<>();
+        private int maxIdx;
+
+        /**
+         * Returns true if cleanable list is empty.
+         *
+         * @return true if the list is empty
+         */
+        public boolean isEmpty() {
+            synchronized (list) {
+                return list.isEmpty();
+            }
+        }
+
+        /**
+         * Insert this PhantomCleanable in the list.
+         */
+        public void insert(PhantomCleanable<?> phc) {
+            synchronized (list) {
+                // Inserting at the end, record the indexes.
+                phc.index = maxIdx = list.size();
+                list.add(phc);
+                assert list.get(phc.index) == phc;
+            }
+        }
+
+        /**
+         * Remove this PhantomCleanable from the list.
+         *
+         * @return true if Cleanable was removed or false if not because
+         * it had already been removed before
+         */
+        public boolean remove(PhantomCleanable<?> phc) {
+            synchronized (list) {
+                int thisIdx = phc.index;
+                if (thisIdx == -1) {
+                    // Not in the list.
+                    return false;
+                }
+
+                // Unlink PhantomCleanable.
+                assert list.get(phc.index) == phc;
+                phc.index = -1;
+
+                int lastIdx = list.size() - 1;
+                if (lastIdx != thisIdx) {
+                    // Move the last, still alive element at current index,
+                    // overwriting the removed one. Update its index to a new location.
+                    PhantomCleanable<?> last = list.get(lastIdx);
+                    last.index = thisIdx;
+                    list.set(thisIdx, last);
+                }
+
+                // Cut the tail. Runs in constant time.
+                list.remove(lastIdx);
+
+                // Capacity control: trim the backing storage if it looks like
```

Comment:
I think we need to set maxIdx to lastIdx - 1 here.  Otherwise, we'll never trim back the storage.  maxIdx can only increase in insert() as currently implemented.

Reply:
Right, d'uh. The unavailability of `ArrayList.capacity()` is what gets us in this mess. I'll try to rewrite to just plain arrays.

Reply:
I pushed the array-based implementation, but ran out of time to properly test it. I'll circle back to it tomorrow. Review if you can, but it will probably have some touchups later.

Reply:
Speaking of testing, with this level of change, perhaps a new (white box?) regtest is warranted.

Reply:
I see that existing `Cleaner` tests are quite rich already. Array-based implementation does rise question about testing the resizes, that's obvious. But I have discovered that copying arrays is not really that good in this scenario which balloons cleaner lists and shrinks them rapidly in our reproducers. So I have reimplemented the whole thing as the linked list of arrays, which makes capacity management code much less problematic. The new asserts around that code should IMO test enough with existing tests.

Reply:
Nevermind, I added a directed regression test for it now.

Reply:
> I see that existing `Cleaner` tests are quite rich already. 
> ...
> The new asserts around that code should IMO test enough with existing tests.

Should any of the existing `Cleaner` tests have `@run`s added with `-esa` ?

Reply:
RunTests.gmk adds -ea -esa so the testing is done with asserts enabled when using "make run-test".

Reply:
So then adding `-ea -esa` is superfluous in the tests. I removed these from the test for extra cleanliness.